### PR TITLE
Add example of a TPM2.0 Quote using wolfTPM

### PR DESCRIPTION
**Pull request by tomoveu**
This PR is a result of #106 enhancement. 

After short fight with a strange TPM error "TPM2_Quote failed 0x3d5: TPM_RC_SENSITIVE: The sensitive area did not unmarshal correctly after decryption", I a happy to share that we now have a generic TPM2_Quote example.

I want to extend the example prints and possibly also output the TPM2_Quote result in a file(blob) for a later use, e.g. verification or other uses. @dgarske  I would appreciate your thoughts on this, because right now the user output of the example is just plain - Confirmation based on the output structure from the TPM2_Quote operation that indeed a PCR has been signed and TPM2 Quote generated.

Additionally, I am starting to work on PCR extend example, so we can demonstrate differences in Quotes. This is where the printing is also important. What do we print to the user who tries our demos.

Signed-off-by: Dimitar Tomov <dimi@designfirst.ee>

### CODE REVIEW

**Comment by tomoveu** [comment]
Also, there are some pcrSelections code that I would like to create helpers for and not be inline. @dgarske let me know where is the right place to create these. Thanks!

File: examples/pcr/quote.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    XMEMCPY(session[0].auth.buffer, usageAuth, session[0].auth.size);
+
+    /* Prepare Quote request */
+    XMEMSET(&cmdIn.quoteAsk, 0, sizeof(cmdIn.quoteAsk));
+    XMEMSET(&cmdOut.quoteResult, 0, sizeof(cmdOut.quoteResult));
+    cmdIn.quoteAsk.signHandle = rsaKey.handle.hndl;
+    cmdIn.quoteAsk.inScheme.scheme = TPM_ALG_RSASSA;
+    cmdIn.quoteAsk.inScheme.details.rsassa.hashAlg = TPM_ALG_SHA256;
+    cmdIn.quoteAsk.qualifyingData.size = 0; /* optional */
+    /* Set the PCR for signing */
```
Comment (dgarske) [inline examples/pcr/quote.c:173]: This section can be replaced with `TPM2_SetupPCRSel(&cmdIn.quoteAsk.PCRselect, TPM_ALG_SHA256, 16);` and `SET_PCR_SELECT_BIT` can be removed.

File: examples/pcr/include.am [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# vim:ft=automake
+# All paths should be given relative to the root
+
+if BUILD_EXAMPLES
+noinst_PROGRAMS += examples/pcr/quote
+noinst_HEADERS  += examples/pcr/quote.h
+examples_pcr_quote_SOURCES      = examples/pcr/quote.c \
+									       examples/tpm_io.c
```
Comment (dgarske) [inline examples/pcr/include.am:8]: Check tabs vs. spaces.

File: examples/README.md [hunk 3]
```diff
@@ -20,14 +20,21 @@ Demonstrates calling the wolfTPM2_* wrapper API's.
 `./examples/wrap/wrap_test`
 
 
-## Attestation Use Case (TPM signed timestamp)
+## Attestation Use Cases
+
+# TPM signed timestamp, TPM2.0 GetTime
```
Comment (dgarske) [inline examples/README.md:25]: Markdown heading should be `###`.

File: examples/README.md [hunk 4]
```diff
@@ -20,14 +20,21 @@ Demonstrates calling the wolfTPM2_* wrapper API's.
...
 
 Demonstrates creation of Attestation Identity Keys(AIK) and the generation of TPM signed timestamp that can be later used as protected report of the current system uptime.
 
 This example demonstrates the use of authSession(authorization Session) and policySession(Policy authorization) to enable the Endorsement Hierarchy necessary for creating AIK. Then, the AIK is used to issue a TPM2_GetTime command using the wolfTPM2 native API. This provides us with TPM generated and signed timestamp that can be used as a system report of its uptime.
 
 `./examples/timestamp/signed_timestamp`
 
+# TPM signed PCR(system) measurement, TPM2.0 Quote
```
Comment (dgarske) [inline examples/README.md:33]: Markdown heading should be `###`.

File: examples/pcr/include.am (hunk 2 above)
Comment (tomoveu) [reply examples/pcr/include.am:8]: Thanks, this is inherited from the signed_timestamp example, which inherited it from the native_tests example.
I will fix all of this in a single commit independently from the functionality, but will put it in this PR.

File: examples/pcr/quote.c (hunk 1 above)
Comment (tomoveu) [reply examples/pcr/quote.c:173]: nice, I was thinking of making such helper and it already exists, dont know how i missed it. thanks, David!

File: examples/README.md (hunk 3 above)
Comment (tomoveu) [reply examples/README.md:25]: Yep, thank you:)

File: examples/README.md (hunk 4 above)
Comment (tomoveu) [reply examples/README.md:33]: yes, thanks

**Comment by tomoveu** [comment]
@dgarske feedback taken in.

I also added debug prints of the PCR digest and Quote signature.

Next, I am adding argument for the user to choose which PCR to quote (PCR16 is default, because it is exactly for debug purposes). Then, I am adding blob output of the TPM2 Quote structure, so we can use it with a verifier(either our example tool or someone else's tool like the user or tpm2_print).

**Comment by tomoveu** [comment]
Thank you, David! I really appreciate the positive feedback.

I added support for parameters of the tool. I kept the arg parsing to a minimum(simple). In wolfssl I noticed there is mygetopt helper, but in wolfTPM it is missing. And considering the current needs I think it would be an overkill to use it. Probably in future, if we have an example with lots of arguments.

**Advanced mode**, with parameters
```
trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/quote 16 test.blob
Demo of generating signed PCR measurement (TPM2.0 Quote)
wolfTPM2_Init: success
wolfTPM2_CreateEK: Endorsement 0x80000000 (314 bytes)
wolfTPM2_CreateSRK: Storage 0x80000001 (282 bytes)
TPM2_StartAuthSession: sessionHandle 0x3000000
wolfTPM2_CreateAndLoadAIK: AIK 0x80000002 (280 bytes)
TPM2_Quote: success
TPM with signature attests (type 0x8018):
	TPM signed 1 count of PCRs
trm@DFDEV2:~/wolfTPM$ tpm2_print -t TPMS_ATTEST -f test.blob
magic: ff544347
type: 8018
qualifiedSigner: 000b15a443a5a17e29d68ce2d4a2135a343a3ecafb9cabd6cda735b8f8a760df2b95
extraData: 
clockInfo:
  clock: 3738918291
  resetCount: 2938180827
  restartCount: 1485897581
  safe: 1
firmwareVersion: 6a49784df76411d2
attested:
  quote:
    pcrSelect:
      count: 1
      pcrSelections:
        0:
          hash: 11 (sha256)
          sizeofSelect: 3
          pcrSelect: 000001
    pcrDigest: 6f20e0f7216b431563868506171423041d4c35231676403c7a79463166936f60
trm@DFDEV2:~/wolfTPM$ 
```

**Demo mode**, without parameters, uses PCR16 and default filename "quote.blob"

```
trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/quote
Demo of generating signed PCR measurement (TPM2.0 Quote)
wolfTPM2_Init: success
wolfTPM2_CreateEK: Endorsement 0x80000000 (314 bytes)
wolfTPM2_CreateSRK: Storage 0x80000001 (282 bytes)
TPM2_StartAuthSession: sessionHandle 0x3000000
wolfTPM2_CreateAndLoadAIK: AIK 0x80000002 (280 bytes)
TPM2_Quote: success
TPM with signature attests (type 0x8018):
	TPM signed 1 count of PCRs
trm@DFDEV2:~/wolfTPM$ tpm2_print -t TPMS_ATTEST -f quote.blob
magic: ff544347
type: 8018
qualifiedSigner: 000b2574e0c86619340cc4a01697eac93583d4bfcc9a05054837ac23006a024c2d4c
extraData: 
clockInfo:
  clock: 3739017287
  resetCount: 449049982
  restartCount: 1188062421
  safe: 1
firmwareVersion: a9e4928c367fa165
attested:
  quote:
    pcrSelect:
      count: 1
      pcrSelections:
        0:
          hash: 11 (sha256)
          sizeofSelect: 3
          pcrSelect: 000001
    pcrDigest: 6f20e0f7216b431563868506171423041d4c35231676403c7a79463166936f60
trm@DFDEV2:~/wolfTPM$ 

```

last, but not least , error handling in case of wrong parameters

```

trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/quote 16
Incorrect arguments
Expected usage: extend (pcr_number) (filename)
* pcr_number is a number between 0-23
* filename for saving the TPMS_ATTEST structure to a file
Demo usage without parameters, generates quote over PCR16 and
saves the output TPMS_ATTEST structure to "quote.blob" file.

```

**Comment by tomoveu** [comment]
Should I add the extend tool to the same PR, it is almost ready, just finishing up the data hashing. Any thoughts on hashing arbitrary data with SHA256 and then passing the digest for PCR Extend?

wolfTPM has no external dependencies , but for the extend we will require wolfcrypt. I guess this is okay?

**Comment by tomoveu** [comment]
@dgarske added the extend tool. It works with arbitrary data from user file. Kept simplistic arguments parsing. 

I am adding a pcr clear tool and then writing the README for the flow that I have in mind as a rich demo. Using PCR_Clear a user could reproduce tests and values(measurements). Making it both easier to understand + useful to experiment with the demo(e.g create and/or test golden values).

I would like your advice on the wolfcrypt dependency of the extend tool. Maybe best to ifdef in the include.am. What would you recommend?

**Comment by tomoveu** [comment]
Some example outputs from the extend helper tool. We can measure files now :-)

With verbose enabled it prints the PCRx value (really useful for debugging)

```
trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/extend 
Demo how to extend data into a PCR (TPM2.0 measurement)
wolfTPM2_Init: success
Hash to be used for measurement:
160000000000000018D54AE9E17F000000000000000000000000000000000000
Command size: 65
...
TPM2_PCR_Extend success
...
PCR16 value:
	91 13 48 c8 e4 7e 9e 2a b5 b1 fb 01 82 85 9e ac | ..H..~.*........
	20 62 f6 ea 45 09 7a ed c5 97 98 8b 76 2e ed 93 |  b..E.z.....v...
...

```

Without debug enabled, it is tight and clear
```

trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/extend 
Demo how to extend data into a PCR (TPM2.0 measurement)
wolfTPM2_Init: success
Hash to be used for measurement:
C0DB240B7D7F0000801F0000FFFF00008824250B7D7F00000000000000000000
TPM2_PCR_Extend success
trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/extend 16 /usr/bin/zipinfo
Demo how to extend data into a PCR (TPM2.0 measurement)
wolfTPM2_Init: success
Hash to be used for measurement:
6C9C165D311DC5490BEAFE23DB0EA3BBD5467CF169C7D63C15B19B91FEF7891D
TPM2_PCR_Extend success
trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/extend 16 /usr/bin/zip
Demo how to extend data into a PCR (TPM2.0 measurement)
wolfTPM2_Init: success
Hash to be used for measurement:
8C8E79CCF51C391E840FBC24E86E1FD261AC80516EBB52F09A28E594FBA25844
TPM2_PCR_Extend success
trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/extend 16 /usr/bin/zipinfo
Demo how to extend data into a PCR (TPM2.0 measurement)
wolfTPM2_Init: success
Hash to be used for measurement:
6C9C165D311DC5490BEAFE23DB0EA3BBD5467CF169C7D63C15B19B91FEF7891D
TPM2_PCR_Extend success
trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/extend 
Demo how to extend data into a PCR (TPM2.0 measurement)
wolfTPM2_Init: success
Hash to be used for measurement:
C04B106C217F0000801F0000FFFF00008894106C217F00000000000000000000
TPM2_PCR_Extend success
trm@DFDEV2:~/wolfTPM$ 

```

Last, but not least, error usage:
```

trm@DFDEV2:~/wolfTPM$ sudo examples/pcr/extend 1
Incorrect arguments
Expected usage:
extend (pcr_number) (filename)
* pcr_number is a PCR index between 0-23
* filename points to file(data) to measure
Demo usage without parameters, extends PCR16 with known hash.

```

**Comment by tomoveu** [comment]
@dgarske  i drove some bicycle and got a better idea :D 

if wolfcrypt is enabled apply calculation of the hash digest over the file data
if NO_WOLFCRYPT take the input data as ready for use SHA256 digest

Reason `--disable-wolfcrypt` has to be EXPLICIT on the `configure` line, so obviously the user has other means of getting what wolfcrypt does, incl. hashes, therefore my assumption is the user has a way to calc hashes OR does not want wolfTPM do perform SW cryptography.

Let me know what do you think 
- disable extend completely if NO_WOLFCRYPT
or
- enable extend even when NO_WOFLCRYPT is enabled, but inform user `printf("Note: Wolfcrypt is not enabled, input data in the file has to be SHA256 Hash digest"

**Comment by dgarske** [comment]
@tomoveu : I like your second option. Let's enable extend with or without wolfCrypt and assume the input file is a computed digest and just take the first 32 bytes of it.

> @dgarske i drove some bicycle and got a better idea :D
> 
> if wolfcrypt is enabled apply calculation of the hash digest over the file data
> if NO_WOLFCRYPT take the input data as ready for use SHA256 digest
> 
> Reason `--disable-wolfcrypt` has to be EXPLICIT on the `configure` line, so obviously the user has other means of getting what wolfcrypt does, incl. hashes, therefore my assumption is the user has a way to calc hashes OR does not want wolfTPM do perform SW cryptography.
> 
> Let me know what do you think
> 
> * disable extend completely if NO_WOLFCRYPT
>   or
> * enable extend even when NO_WOFLCRYPT is enabled, but inform user `printf("Note: Wolfcrypt is not enabled, input data in the file has to be SHA256 Hash digest"

File: examples/pcr/extend.c [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+#include <examples/tpm_test.h>
+
+#include <stdio.h>
+
+
+/******************************************************************************/
+/* --- BEGIN TPM2.0 PCR Extend example tool  -- */
+/******************************************************************************/
+
+inline static void usage(void)
```
Comment (dgarske) [inline examples/pcr/extend.c:38]: Let's just make this `static` and remove inline (better portability).

File: examples/pcr/extend.c [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+{
+    printf("Expected usage:\n");
+    printf("extend (pcr_number) (filename)\n");
```
Comment (dgarske) [inline examples/pcr/extend.c:41]: Better usage template: `./examples/pcr/extend [pcr_number] [filename]`. Also add `(default 16) in the `pcr_number` description.

File: examples/pcr/extend.c [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+{
+    int i, pcrIndex, len, rc = -1;
+    WOLFTPM2_DEV dev;
+    /* Arbitrary user data provided through a file */
+    unsigned char dataBuffer[1024];
+    const char *filename = NULL;
+    FILE *dataFile = NULL;
+    /* Using wolfcrypt to hash input data */
+    unsigned char hash[SHA256_DIGEST_SIZE];
+    Sha256 sha256;
```
Comment (dgarske) [inline examples/pcr/extend.c:57]: As discussed you'll need to support `WOLFTPM2_NO_WOLFCRYPT` case and assume provided file is raw hash digest.

File: examples/pcr/extend.c [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        pcrIndex = strtol(argv[1], NULL, 10);
+        if(pcrIndex < 0 || pcrIndex > 23) {
+            printf("PCR index is out of range(0-23)\n");
+            goto exit_badargs;
+        }
+
+        filename = argv[2];
+        dataFile = fopen(filename, "rb");
+        if(dataFile == NULL) {
+            printf("Error openning file %s\n", filename);
```
Comment (dgarske) [inline examples/pcr/extend.c:88]: Spell `opening`

**Comment by tomoveu** [comment]
@dgarske all changes made + NO_WOLFCRYPT support as discussed

Remains to add the clear helper tool and script to showcase the complete demo + README

ps: Only FYI, cause I was caught by surprise :D I fixed one corner issue with the printf of the hash to be used(important print) and also one missing header when NO_WOLFCRYPT was enabled. Slipped by the review ;)

**Comment by tomoveu** [comment]
OK, just added the PCR Reset tool. Bit later will create a nice README + script to showcase everything.

File: .gitignore [hunk 9]
```diff
@@ -34,6 +34,7 @@ examples/csr/csr
 examples/tls/tls_client
 examples/pkcs7/pkcs7
 examples/timestamp/signed_timestamp
+examples/timestamp/pcr
```
Comment (dgarske) [inline .gitignore:37]: Looks like this should be:

```
examples/pcr/extend
examples/pcr/quote
examples/pcr/reset
```

File: examples/README.md [hunk 10]
```diff
@@ -20,14 +20,23 @@ Demonstrates calling the wolfTPM2_* wrapper API's.
...
 
 `./examples/timestamp/signed_timestamp`
 
+### TPM signed PCR(system) measurement, TPM2.0 Quote
+
+Demonstrates the generation of TPM2.0 Quote used for attestation of the system state by putting PCR value(s) in a TPM signed structure.
+
+More information about how to test and use PCR attestation can be found in the in README file located in the `pcr` folder of the example.
+
+`./examples/pcr/quote`
```
Comment (dgarske) [inline examples/README.md:39]: TODO: `examples/pcr/extend` and `examples/pcr/reset`

File: examples/pcr/extend.c [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+int TPM2_Extend_Test(void* userCtx, int argc, char *argv[])
+{
+    int i, pcrIndex, len, rc = -1;
+    WOLFTPM2_DEV dev;
+    /* Arbitrary user data provided through a file */
+    const char *filename = NULL;
+    FILE *dataFile = NULL;
+    unsigned char hash[TPM_SHA256_DIGEST_SIZE];
```
Comment (dgarske) [inline examples/pcr/extend.c:61]: For consistency you could use `byte` here. It's always available and used for all wolf libs.

**Comment by tomoveu** [comment]
@dgarske feedback taken in. Tomorrow I will push the script and special README for the quote demo involving all new examples.

**Comment by tomoveu** [comment]
@dgarske README and script pushed. 

I had several variations and had to force push the right one. Script is simple, value is in the README. I think this one has the most clear explanation.

**Comment by tomoveu** [comment]
Thank you @embhorn ! I really appreciate the feedback.

Thank you @dgarske for the finishing touches. I am so glad you think the examples are excellent.

Only one technical note, I used strtol, because atoi is considered obsolete and "0" is difficult to detect - https://www.gnu.org/software/libc/manual/html_node/Parsing-of-Integers.html Either way, atoi or strtol are not the focus here and I am so happy we now have TPM2 Quote example! 💯

**Pull request merged**